### PR TITLE
Expose and accept version coauthor information

### PR DESCRIPTION
**Pull request by fashxp**
## Summary

Exposes the new core version coauthor fields (`coauthorType`, `coauthor`) in the versions API and lets clients set them on element saves and on existing versions.

Core dependency pimcore/pimcore#19235 is **merged** (2026-07-02), so this is no longer blocked.

## Changes

- `Version` listing schema + `VersionHydrator`: optional nullable `coauthorType`/`coauthor` (NOT added to `required` — additive, BC-safe).
- Save/patch payloads accept optional `coauthorType`/`coauthor` keys (same pattern as `task`). `UpdateService`/`PatchService` delegate to the new `Element\Service\CoauthorService`, which validates the pair and runs the save inside core's `CoauthorContextInterface::withCoauthor()` scope, so the previous context is restored exception-safely by core rather than by hand-rolled snapshot/restore in each service.
- Overlong values are rejected at the API boundary instead of being truncated by the save: the widths of the core `versions` columns (50/255) live in `Util\Constant\VersionCoauthor` and are enforced as a 422 on the update/patch payloads and as a `Length` constraint on `PUT /versions/{id}`. They are surfaced in OpenAPI through the new `CoauthorTypeProperty`/`CoauthorProperty` attributes, which also de-duplicate the five request bodies.
- `PUT /versions/{id}` accepts optional `coauthorType`/`coauthor` (like `note`/`public`); empty string clears the field.
- New typed constants on `ElementSaveServiceInterface`; OpenAPI `Property` declarations on the five update/patch request bodies.

## Verification

- Full Codeception Unit suite green (503 tests, 1150 assertions), PHPStan clean, php-cs-fixer clean.
- New tests: `PatchServiceTest` (9 — context activation, restoration after success and after an exception, outer-context restoration, no-keys/empty-string cases, length limits and boundary values), `UpdateServiceTest` (+2 length cases), `UpdateVersionParameterTest` (5 — the `PUT` length constraints), plus the existing `VersionHydratorTest`.
- Live-verified on the enterprise demo: payload-key save stamped the version row, `GET /versions/{elementType}/{id}` returns the fields, `PUT /versions/{id}` sets and clears them.

### Known gap

`VersionRepository::updateVersion()` has no unit test. `Pimcore\Model\Version` is `final` so it cannot be mocked, and instantiating a real one outside a booted kernel is not possible in the Unit suite, so that seam remains covered by the live verification above.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01SsEa2MTm2gRPyGEdHGnnoF

### CODE REVIEW

**Comment by fashxp** [comment]
CI status confirmed: all failing checks (PHPStan matrix, Codeception matrix, Sonar analysis) fail solely on `unknown class Pimcore\Model\Version\CoauthorContextInterface` — CI installs `pimcore/pimcore` from Packagist, which does not yet contain pimcore/pimcore#19235. No failures originate from this PR's own code (local: full Unit suite green, PHPStan clean, cs-fixer clean). Re-run after the core PR merges.

**Comment by fashxp** [comment]
@copilot resolve the merge conflicts in this pull request

File: src/Updater/Service/UpdateService.php [hunk 1]
```diff
@@ -38,7 +38,12 @@
 use Pimcore\Model\Element\DuplicateFullPathException;
 use Pimcore\Model\Element\ElementInterface;
 use Pimcore\Model\Element\ValidationException;
+<<<<<<< HEAD
+use Pimcore\Model\Version\CoauthorContextInterface;
+use function is_string;
+=======
 use function in_array;
+>>>>>>> origin/2026.x
```
[the remark this answers is not in the document]
Comment (fashxp) [reply src/Updater/Service/UpdateService.php]: Resolved in e2e1952c. Both conflicts kept all additions from either side; `php -l`, PHPStan and the Unit suite are green again.

File: src/Patcher/Service/PatchService.php [hunk 2]
```diff
@@ -66,7 +68,11 @@ public function __construct(
         private JobExecutionAgentInterface $jobExecutionAgent,
         private ElementIndexServiceInterface $indexService,
         private ElementSaveServiceInterface $elementSaveService,
+<<<<<<< HEAD
+        private CoauthorContextInterface $coauthorContext,
+=======
         private SecurityServiceInterface $securityService
+>>>>>>> origin/2026.x
```
[the remark this answers is not in the document]
Comment (fashxp) [reply src/Patcher/Service/PatchService.php]: Resolved in e2e1952c. Both constructor dependencies are kept, and in `patchElement()` the permission checks now run before the coauthor context is set up so a `ForbiddenException` cannot leave context state behind.

File: composer.json [hunk 3]
```diff
@@ -22,7 +22,7 @@
     "nesbot/carbon": "^3.10.0",
     "pimcore/static-resolver-bundle": "^2026.2.0",
     "pimcore/generic-data-index-bundle": "^2026.2.0",
-    "pimcore/pimcore": "^2026.2.0",
+    "pimcore/pimcore": "^2026.3",
```
[the remark this answers is not in the document]
Comment (fashxp) [reply composer.json:25]: Not applying this one. The bump is required and matches how this repo handles it: #1912 ("[User] Add theme attribute to User") bumped `pimcore/pimcore` from `^2026.1.5` to `^2026.2.0` in exactly the same situation, a feature PR exposing a new core field through the Studio API. The constraint resolves via the dev dependency and then against 2026.x; CI at 563feadc was fully green with it, including the legs that install from Packagist.

File: src/Version/MappedParameter/UpdateVersionParameter.php [hunk 4]
```diff
@@ -21,6 +21,8 @@
     public function __construct(
         private ?bool $public = null,
         private ?string $note = null,
+        private ?string $coauthorType = null,
+        private ?string $coauthor = null,
```
[the remark this answers is not in the document]
Comment (fashxp) [reply src/Version/MappedParameter/UpdateVersionParameter.php:30]: Added in e2e1952c: `#[Length(max: VersionCoauthor::MAX_TYPE_LENGTH)]` / `MAX_COAUTHOR_LENGTH` on the parameter, with `maxLength` mirrored into the OpenAPI schema. Covered by the new `UpdateVersionParameterTest`.

File: src/Updater/Service/UpdateService.php [hunk 5]
```diff
@@ -109,10 +117,48 @@ public function update(string $elementType, int $id, array $data): void
...
+<<<<<<< HEAD
+     * @return array{type: ?string, coauthor: ?string}|null Snapshot of the previous context,
+     *                                                      or null if nothing was activated
+     */
+    private function activateCoauthorContext(array $data): ?array
+    {
+        $coauthorType = $data[ElementSaveServiceInterface::INDEX_COAUTHOR_TYPE] ?? null;
+        $coauthor = $data[ElementSaveServiceInterface::INDEX_COAUTHOR] ?? null;
+
+        if (!is_string($coauthorType) || $coauthorType === '' || !is_string($coauthor) || $coauthor === '') {
```
[the remark this answers is not in the document]
Comment (fashxp) [reply src/Updater/Service/UpdateService.php]: Added in e2e1952c. `CoauthorContextTrait::extractPayloadCoauthor()` validates both limits and throws `InvalidArgumentException` (422). It runs before the save try/catch on purpose - inside it, the exception would be swallowed by `catch (Exception)` and resurface as a 500.

File: src/Patcher/Service/PatchService.php [hunk 6]
```diff
@@ -189,9 +199,48 @@ public function patchElement(
...
+    /**
+     * @return array{type: ?string, coauthor: ?string}|null Snapshot of the previous context,
+     *                                                      or null if nothing was activated
+     */
+    private function activateCoauthorContext(array $elementPatchData): ?array
+    {
+        $coauthorType = $elementPatchData[ElementSaveServiceInterface::INDEX_COAUTHOR_TYPE] ?? null;
+        $coauthor = $elementPatchData[ElementSaveServiceInterface::INDEX_COAUTHOR] ?? null;
+
+        if (!is_string($coauthorType) || $coauthorType === '' || !is_string($coauthor) || $coauthor === '') {
```
[the remark this answers is not in the document]
Comment (fashxp) [reply src/Patcher/Service/PatchService.php]: Same fix as on the update path - the shared trait validates both limits before the patch save runs. Covered by `PatchServiceTest::testPatchRejectsOverlongCoauthorType/CoauthorBeforeSaving`.

File: src/Patcher/Service/PatchService.php [hunk 7]
```diff
@@ -189,9 +199,48 @@ public function patchElement(
             throw $exception;
         } catch (Exception $exception) {
             throw new ElementSavingFailedException($element->getId(), $exception->getMessage());
+        } finally {
+            $this->restoreCoauthorContext($coauthorSnapshot);
```
[the remark this answers is not in the document]
Comment (fashxp) [reply src/Patcher/Service/PatchService.php]: Added in e2e1952c: `tests/Unit/Patcher/Service/PatchServiceTest.php` covers activation during save, restoration after success and after an exception, restoration of an outer context, the no-keys and empty-string cases, the length limits and the boundary values - 9 tests.

File: src/Version/Repository/VersionRepository.php [hunk 8]
```diff
@@ -135,6 +135,14 @@ public function updateVersion(
             $version->setPublic($parameter->isPublic());
         }
 
+        if ($parameter->getCoauthorType() !== null) {
+            $version->setCoauthorType($parameter->getCoauthorType() === '' ? null : $parameter->getCoauthorType());
+        }
+
+        if ($parameter->getCoauthor() !== null) {
+            $version->setCoauthor($parameter->getCoauthor() === '' ? null : $parameter->getCoauthor());
```
[the remark this answers is not in the document]
Comment (fashxp) [reply src/Version/Repository/VersionRepository.php:143]: Not covered by a unit test, and I do not think it can be in this suite: `Pimcore\Model\Version` is `final` (models/Version.php:39) so it cannot be mocked, and instantiating a real one outside a booted Pimcore kernel crashes the process, so `updateVersion()` cannot be exercised without a database. The bundle has no DB-backed repository test to follow either - `tests/Unit` currently contains no repository tests at all.

What I did instead is cover the part of this boundary that is testable: `UpdateVersionParameterTest` covers the new length constraints, including empty strings staying valid because they are the clear signal. The set/partial-update/empty-string-clear behaviour in `updateVersion()` itself remains covered only by the live verification noted in the PR description. Happy to add a functional test if you want one, but it would need a suite that boots the kernel.

**Review by lukmzig — APPROVED** [review]
